dev/show_package_release_dates.py

```diff
@@ -0,0 +1,79 @@
+import subprocess
+import requests
+from concurrent.futures import ThreadPoolExecutor
+import traceback
+
+
+def get_distributions():
+    res = subprocess.run(["pip", "list"], stdout=subprocess.PIPE)
+    pip_list_stdout = res.stdout.decode("utf-8")
+    # `pip_list_stdout` looks like this:
+    # ``````````````````````````````````````````````````````````
+    # Package     Version             Location
+    # ----------- ------------------- --------------------------
+    # mlflow      1.21.1.dev0         /home/user/path/to/mlflow
+    # tensorflow  2.6.0
+    # ...
+    # ``````````````````````````````````````````````````````````
+    lines = pip_list_stdout.splitlines()[2:]  # `[2:]` removes the header
+    return [
+        # Extract package and version
+        line.split()[:2]
+        for line in lines
+    ]
+
+
+def get_release_date(package, version):
+    resp = requests.get(f"https://pypi.python.org/pypi/{package}/json")
+    if not resp.ok:
+        return ""
+
+    matched = [dist_files for ver, dist_files in resp.json()["releases"].items() if ver == version]
+    if (not matched) or (not matched[0]):
+        return ""
+
+    upload_time = matched[0][0]["upload_time"]
+    return upload_time.split("T")[0]  # return year-month-day
+
+
+def get_longest_string_length(array):
+    return len(max(array, key=len))
+
+
+def safe_result(future, if_error=""):
+    try:
+        return future.result()
+    except Exception:
+        traceback.print_exc()
+        return if_error
+
+
+def main():
+    distributions = get_distributions()
+    with ThreadPoolExecutor() as executor:
```

Comment:
Default max_workers value of os.cpu_count() * 5 is a 'goldilock zone' default that strikes a balance between moderate CPU-bound tasks and I/O bound tasks. Py3.8 `concurrent.futures.ThreadPoolExecutor` did a config change for this default that might be helpful in the case of spotty socket connection to pypy, reserving a few threads for longer response times without incurring excessive context switching. 
```python
with ThreadPoolExecutor(max_workers=min(32, os.cpu_count() + 4)):
```
Just a thought.

Reply:
Thanks for the knowledge! I just read https://bugs.python.org/issue35279. Using `min(32, os.cpu_count() + 4)` sounds good to me.